### PR TITLE
Explicit imports now shadow locally defined identifiers

### LINKED ISSUE #9552: identical code compiles in separate source files, fails in single source file
As a single file, this doesn't compile:
```scala
package object p {
  type Foo = Int
}
package p {
  object Dingo { type Foo = String }
  trait Bippy  {
    import Dingo._
    def z: Foo
    z: String
  }
}
// s1/a.scala:8: error: reference to Foo is ambiguous;
// it is both defined in package object p and imported subsequently by
// import Dingo._
//     def z: Foo
//            ^
// one error found
```
However if that code is placed in two separate files, one for the package object and one for the rest, it compiles. I emphasize this is not a separate compilation bug - in both cases all the source code is being given to the compiler. The ONLY difference is the number of source files.

There is an additional problem. If we move import Dingo._ up one line so it is outside of Bippy, then the two-file variation no longer compiles, but not due to identifier ambiguity!
```scala
s2/b.scala:6: error: type mismatch;
 found   : p.Foo
    (which expands to)  Int
 required: String
    z: String
    ^
one error found
```
Notice it goes from resolving Foo one way to resolving Foo the other way after the minimum possible change. Sanity suggests the minimum change must be rendered ambiguous at one point or the other, as of course it is in the single-file case.

### PR DESCRIPTION
**Pull request by som-snytt**
Original title: "Import shadows def in other unit, per spec"

Forward port of #5339 .

Fixes scala/bug#2458
Fixes scala/bug#9552

### CODE REVIEW

**Comment by som-snytt** [comment]
NB: two import fixes were still required, two others in tests pass after "Forgive import ambiguity". Those changes squashed in lead commit.

**Comment by som-snytt** [comment]
Even Seth doesn't know the password.

Hey, how can settings be broken already? I'm only creating a `-Y` option that forks language semantics.

Edit: rebased

**Comment by SethTisue** [comment]
/synch goddamn it Scabot or I'll come over and put my foot right through your circuits

**Comment by hrhino** [comment]
"important" is clever but the previous PR had the benefit of ~~iambic~~ trochaic pentameter.

**Comment by som-snytt** [comment]
Maybe I broke cla. Is Lightbend CLA different? Is it possible I'm in one system but not the other?

https://github.com/sbt/sbt/pull/4143#issuecomment-386659744

**Comment by som-snytt** [comment]
Sample erroneous shadowing. The `import java.util` isn't used, of course.
```
[warn] /home/amarki/projects/scala/test/junit/scala/tools/nsc/DeterminismTest.scala:5: imported `util' is permanently hidden by definition of package util in package nsc
```

File: src/reflect/scala/reflect/internal/TreeInfo.scala [hunk 1]
```diff
@@ -861,23 +861,32 @@ abstract class TreeInfo {
...
+      case _                            => false
     }
-
-    isUnitInScala(body, nme.Predef) || isLeadingPredefImport(body)
+    def goodPreambleImport(sym: Symbol): Boolean = {
+      val ok = sym.hasPackageFlag || {
+        val pkg = sym.enclosingPackage.name
+        !unitDefinesClassOrObject(pkg, sym.name) && !hasExplicitImport(body, pkg, sym.name)
+      }
+      if (!ok) report(sym)
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/TreeInfo.scala]: https://xkcd.com/838/

**Review by lrytz — CHANGES_REQUESTED** [review]
I think we should split out the `-Yimports` change into a separate PR.

The changes should come with tests.

Could you also summarize the changes in the commit messages?

What's the reason for changing the imports (the second commit)?

**Comment by som-snytt** [comment]
I was hoping for feedback on simple `-Yimports` versus the other proposal, but I think the folks who want the mechanism are just using Haskell.

The imports in the few changed files are ambiguous.

Very well, I'll split out `-Yimports`; I don't think the code requires working imports, but using them correctly does. I think that these two-year-old PRs are more likely to die on the vine again.

The linked PR on 2.12 has full commit messages; I didn't expect it to get shut down.

**Comment by som-snytt** [comment]
Also, the other PR includes the spec change for the precedence of "root" imports. As the message here says, this was a minimal forward port.

**Comment by lrytz** [comment]
I definitely like what I saw from `-Yimports`, but I didn't test it locally to see what it does in detail. Reverse-engineering all corner cases from looking at the code was a bit too hard, so either test cases or more description would help discussing the proposal.

> The imports in the few changed files are ambiguous.

It confused me that the commit fixing imports comes after the change to imports handling. But now I see that's a case of https://help.github.com/articles/why-are-my-commits-in-the-wrong-order

**Comment by som-snytt** [comment]
That just happened to me on another PR. I'll rebase and recreate commits here, adding what's missing from the 2.12 PR. Thanks for taking a look. I have high hopes for `-Yimports`, especially as a way for newbies to auto-import implicits, such as `ExecutionContext.Implicits` etc.

**Review by lrytz — APPROVED** [review]

**Comment by sjrd** [comment]
I have added the `release-notes` label to this PR, because it definitely breaks code. I've a bunch of code broken by this change in Scala.js.

**Comment by som-snytt** [comment]
@sjrd I expected there might by a community build to sanity check the nuances. There were two tests that needed updates; and there might be useful idioms that need a pass.

**Comment by sjrd** [comment]
The community build for 2.13.x is virtually empty still, because the ecosystem hasn't been able to catch up with the new collections. So we're still completely blind regarding such changes.

**Comment by SethTisue** [comment]
> The community build for 2.13.x is virtually empty still, because the ecosystem hasn't been able to catch up with the new collections. So we're still completely blind regarding such changes

22 projects are green currently, but that's only 1/8 of the total.

I expect that situation to start improving quickly pretty soon, now that specs2 and ScalaTest are both unblocked (by ScalaCheck having published) and working on support.

I will update the community build right now to use a SHA that includes this PR, so we'll at least start getting feedback on it.

**Comment by som-snytt** [comment]
BTW, I didn't mean to impugn the community build; I realize there are other priorities and things to do. I just meant to say that I want this PR to be more helpful than disruptive, so I'll be alert to either help out or proactively try some builds.

I see the scala-js changes are due to respecting wild card imports `scala.tools.nsc._` for `util` and `scala.collection._` for `Iterable` and friends. In that case, I would have corrected the imports, `scala.collection.{Iterable => _, _}` and so on, rather than qualifying `js.Iterable` everywhere.

I wonder if a scalafix is possible, he said fixedly.

**Comment by SethTisue** [comment]
> I didn't mean to impugn the community build

no offense taken :-)

**Comment by SethTisue** [comment]
I'm certainly seeing other projects in the community build needing changes because of this.. specs2 and ScalaTest, for two. They are similar to the changes described in Som's last comment — wildcard imports now pulling in identifiers that change the meaning of subsequent imports.

**Comment by som-snytt** [comment]
@SethTisue is there a quickie link to the build? so I can see what is failing? thx.

I realize it may take a few minutes for you to notice that your phone is buzzing with a github notification; those swanky bistros can get noisy around brunch time.

**Comment by SethTisue** [comment]
for specs2, check out the master branch and do `++2.13.0-pre-1a172fd!` then `set every scalaBinaryVersion := "2.13.0-M4"` then `coreJVM/Test/compile`

for scalatest, check out the `community-build-2.13` branch of the scalacommunitybuild fork, then `++2.13.0-pre-1a172fd`, `set every scalaBinaryVersion := "2.13.0-M4"`, `scalatest/test:compile`

**Comment by som-snytt** [comment]
I never know what to try a build against. Scalatest has a branch called `funkiness`, which sounds like something I should build

**Comment by som-snytt** [comment]
OK, thanks @SethTisue and jinx.

**Comment by som-snytt** [comment]
I didn't understand `scalacommunitybuild fork`, so I guess I'll try specs2, which I assumed to be harder.

**Comment by som-snytt** [comment]
This is why Seth is Seth.
```
sbt.librarymanagement.ResolveException: download failed: org.kohsuke#github-api;1.92!github-api.jar
```

The other day, my Java friend was struggling to use jaxb (!) when I said I forgot Kohsuke worked on that stuff.

And I can't get the specs2 build to download the dependency either. I don't know the trailing `!` syntax, but I tried w/w/out `!` or `SNAPSHOT`.

```
[error] sbt.librarymanagement.ResolveException: unresolved dependency: org.scala-lang#scala-library;2.13.0-pre-1a172fd: not found
```

I don't know why anyone coined the term "dependency hell."

Wait, did I just spend an hour on that? I just want to run a build. I don't even want to restarr anything.

**Comment by SethTisue** [comment]
> I didn't understand scalacommunitybuild fork

https://github.com/scalacommunitybuild/scalatest/tree/community-build-2.13

> I don't know the trailing ! syntax

sbt 1 changed `++` to only operate on subprojects that already have the specified version listed in `crossScalaVersions`, but you can override that with `!`, a syntax that 0.13 doesn't recognize. ScalaTest is built with sbt 0.13, specs with sbt 1

to get `2.13.0-pre-1a172fd` you need `resolvers += "scala-integration" at "https://scala-ci.typesafe.com/artifactory/scala-integration/"`

I don't know, offhand, why `org.kohsuke#github-api` would fail to resolve for you.

a recent 2.13 community build log is https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1182/consoleFull

**Comment by som-snytt** [comment]
@SethTisue thanks for your help, Seth. I'll follow up, while I'm watching the Secret Circle on CW Seed.

The guy who had a role on Backstrom is my favorite. He must be cursed to have good roles on programs that don't survive their first season. I don't remember him on Heroes but he was also in Terminator, which I skipped, even though like everyone I'm a fan of Summer Glau.

**Comment by som-snytt** [comment]
Sorry to be lame. Why is there `git blame` but not `git lame`? Thought I'd clear up sbt-io:

```
sbt:IO Root> 
sbt:IO Root> set resolvers += "pr" at "https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/"
[info] Defining resolvers
[info] The new value will be used by externalResolvers
[info] Reapplying settings...
[info] Set current project to IO Root (in build file:/Users/andrew/workspace/sbt-io/)
sbt:IO Root> ++2.13.0-pre-1a172fd!
[info] Forcing Scala version to 2.13.0-pre-1a172fd on all projects.
[info] Reapplying settings...
[info] Set current project to IO Root (in build file:/Users/andrew/workspace/sbt-io/)
sbt:IO Root> set every scalaBinaryVersion := "2.13.0-M4"
[info] Defining Global / scalaBinaryVersion, ThisBuild / scalaBinaryVersion and 3 others.
[info] The new values will be used by Compile / compileAnalysisFilename, Compile / packageBin / artifactPath and 37 others.
[info] 	Run `last` for details.
[info] Reapplying settings...
[info] Set current project to IO Root (in build file:/Users/andrew/workspace/sbt-io/)
sbt:IO Root> 
```
but that fails to resolve the compiler artifacts. Probably I'm doing something basically wrong.
```
org.scala-lang#scala-library;2.13.0-pre-1a172fd: not found
```

It's a breezy day here and the little bistro table we got for the patio with an umbrella just blew over. So I just came out to sit in the sun. The heavy chimes from Costco which never chime are chiming. The weird grassy thing that grew last year that looks like corn as in maize which I love but the spouse doesn't, stares me down. Dare I transplant it to outside the fence?

You can tell I've been reading Wallace Stevens lately.

**Comment by sjrd** [comment]
Try
```scala
set resolvers in Global += ...
```
That's what I always use. It's necessary if there are multiple projects in the build.

**Comment by som-snytt** [comment]
@sjrd thanks, I tried and failed. I'll try again. Currently, Jake is back in town on Secret Circle. He's the witch who completes the circle since they killed his brother. You'd think that if I can follow The Secret Circle, I could follow sbt.

**Comment by SethTisue** [comment]
ScalaTest diffs: https://github.com/scalacommunitybuild/scalatest/commit/448a22c7c5f59591042e7df6b749bc0a8321c94c (maybe not stylistically the best, perhaps I should have made the `import org.scalactic._` changes go in the opt-in direction rather than the `FooBar => _` opt-out direction I chose... idk)

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 2]
```diff
@@ -1250,12 +1273,18 @@ trait Contexts { self: Analyzer =>
       }
     }
 
-  } //class Context
+    def isPackageOwnedInDifferentUnit(s: Symbol): Boolean =
+      if (s.isOverloaded) s.alternatives.exists(isPackageOwnedInDifferentUnit)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala:1277]: How can a package-owned symbol be overloaded?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala:1277]: package object?

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala:1277]: Good point! Hadn't considered those... FTR, had to convince myself in the REPL:

```
scala> val s = typeOf[scala.Error].typeSymbolDirect // type alias defined in the scala package object
s: $r.intp.global.Symbol = type Error

scala> s.isDefinedInPackage
res24: Boolean = true

scala> s.effectiveOwner
res26: $r.intp.global.Symbol = package scala

scala> s.effectiveOwner.hasPackageFlag
res27: Boolean = true

scala> s.owner
res28: $r.intp.global.Symbol = package object scala

scala> s.owner.hasPackageFlag
res29: Boolean = false
```

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: TFTRR! (thanks for the repl record) I was about to try it as well, but distracted with backporting the "long result printing fix" for hrhino. The REPL looks so different way back in 2.12!

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Contexts.scala]: I also have to forward port a few niceties, inevitably.

**Comment by SethTisue** [comment]
[specs2 diffs](https://github.com/etorreborre/specs2/pull/680) — this one was easier

**Comment by som-snytt** [comment]
@SethTisue thx, I'll take a look. I would hope the errors would be clarifying. While making supper, I was singing Hindemith's settings of Whitman's engagement with the Civil War. They are suddenly moving to me, I couldn't restrain the tears for our time. The spouse asked if I was sick. Yes! Yes, sick! She grew up under her namesake Mao, so the situation in her homeland looks quite normal, even benign, but I grew up during the Vietnam war and Nixon, and as Whitman said in Hindemith's music, die Schlaefer sollen in diesen Tagen nicht schlafen. I thought I had been radicalized by my radical girlfriend, or my grad school training, but in fact I have not yet woken to the facts of the world. We must leave off thoughts of rest or retirement. We must not sleep. We must speak the words that are most true. No other words will do. Sneaky ell, how word becomes a world. The way 1l is either one or ten more.

**Comment by SethTisue** [comment]
There is now some further discussion and justification at https://github.com/scala/bug/issues/11593

**Pull request merged**